cmd/sma.go

```diff
@@ -0,0 +1,79 @@
+package cmd
+
+import (
+	"sync"
+	"time"
+)
+
+type movingAverage struct {
+	mu sync.Mutex
+
+	window int // Simple Moving Average window (number of latest values to calculate average value)
+
+	vals []float64 // stored values, circular list
+	n    int       // number of actually stored values (vals slice usage)
+	i    int       // last-set value index; first-set (oldest) value is (i+1)%window when n == window
+
+	avg float64 // current Simple Moving Average value
+
+	frequency  time.Duration
+	lastUpdate time.Time
+}
+
+// NewMovingAverage constructs new Simple Moving Average calculator.
+// Window arg must be >= 1.
+func newMovingAverage(window int, frequency time.Duration) movingAverage {
+	if window <= 0 {
+		panic("movavg.NewSMA: window should be > 0")
+	}
+	if frequency == 0 {
+		frequency = time.Second
+	}
+	return movingAverage{
+		window:    window,
+		frequency: frequency,
+		vals:      make([]float64, window),
+	}
+}
+
+// Add recalculates Simple Moving Average value and returns it.
+func (a *movingAverage) Add(v float64) {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	now := time.Now()
```

Comment:
Right now in the critical path, the collection of time taken involves 3 sys calls to get the current time. We should only need 2. The critical path needs to know total time taken by the API call and count (and min/max) and another thread should calculate the moving average and clear the values. You can implement an exponential moving average as well to smooth things out.

Reply:
> Right now in the critical path, the collection of time taken involves 3 sys calls to get the current time. We should only need 2. The critical path needs to know total time taken by the API call and count (and min/max) and another thread should calculate the moving average and clear the values. You can implement an exponential moving average as well to smooth things out.

Yes will do.